Testing.Rmd

```diff
@@ -207,210 +310,434 @@ A test failure looks something like this:
+  snapshot.
+* `variant`: Sometimes snapshots reflect the ambient conditions, such as the
+  operating system or the version of R or one of your dependencies, and you need
+  a different snapshot for each variant. This is an experimental and somewhat
+  advanced feature, so if you can arrange things to use a single snapshot, you
+  probably should.
+  
+In typical usage, testthat will take care of managing the snapshot files below `tests/testthat/_snaps/`.
+This happens in the normal course of you running your tests and, perhaps, calling `testthat::snapshot_accept()`.
 
-When you use these actions in tests, you'll need to clean up after yourself.
-While many other testing packages have set-up and teardown methods that are run automatically before and after each test, these are not so important with testthat because you can create objects outside of the tests and you can rely on R's copy-on-modify semantics to keep them unchanged between test runs.
-To clean up other actions you can use regular R functions.
+### Shortcuts for other common patterns
 
-### What to test
+We conclude this section with a few more expectations that come up frequently.
+But remember that testthat has [many more pre-built expectations](https://testthat.r-lib.org/reference/index.html) than we can demonstrate here.
+
+Several expectations can be described as "shortcuts", i.e. they streamline a pattern that comes up often enough to deserve its own wrapper.
+
+* `expect_match(object, regexp, ...)` is a shortcut that wraps
+  `grepl(pattern = regexp, x = object, ...)`.
+  It matches a character vector input against a regular expression `regexp`.
+  The optional `all` argument controls whether all elements or just one element
+  needs to match.
+  Read the `expect_match()` documentation to see how additional arguments, like
+  `ignore.case = FALSE` or `fixed = TRUE`, can be passed down to `grepl()`.
+   
+    ```{r, error = TRUE}
+    string <- "Testing is fun!"
+  
+    expect_match(string, "Testing") 
+     
+    # Fails, match is case-sensitive
+    expect_match(string, "testing")
+  
+    # Passes because additional arguments are passed to grepl():
+    expect_match(string, "testing", ignore.case = TRUE)
+    ```
+* `expect_length(object, n)` is a shortcut for
+  `expect_equal(length(object), n)`.
+* `expect_setequal(x, y)` tests that every element of `x` occurs in `y`, and
+  that every element of `y` occurs in `x`.
+  But it won't fail if `x` and `y` happen to have their elements in a different
+  order.
+* `expect_s3_class()` and `expect_s4_class()` check that an object `inherit()`s
+  from a specified class. `expect_type()`checks the `typeof()` an object.
+
+    ```{r, error = TRUE}
+    model <- lm(mpg ~ wt, data = mtcars)
+    expect_s3_class(model, "lm")
+    expect_s3_class(model, "glm")
+    ```
+
+`expect_true()` and `expect_false()` are useful catchalls if none of the other expectations does what you need.
+
+## What to test
```

Comment:
below there is some "I" in an item

Reply:
reg writing test when there is a bug, maybe add a footnote about reprex, not really on topic but I was thinking about the bliss when an user gives a reprex, you put it in a test, you fix the code then close the issue